### PR TITLE
Deprecate `PrecisionPlugin.on_save/load_checkpoint`

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Part of #11824
Deprecate `PrecisionPlugin.on_save/load_checkpoint` hooks

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 1]
```diff
@@ -94,20 +94,17 @@ def optimizer_step(
         return closure_result
 
     def state_dict(self) -> Dict[str, Any]:
+        """``ApexMixedPrecisionPlugin.on_save_checkpoint`` is deprecated in v1.6.
+
+        Lightning will auto-save ApexMixedPrecisionPlugin state with ``ApexMixedPrecisionPlugin.state_dict`` instead
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/apex_amp.py:99]:
```suggestion
        Lightning will auto-save ApexMixedPrecisionPlugin state with ``ApexMixedPrecisionPlugin.state_dict`` instead.
```

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 2]
```diff
@@ -109,23 +109,19 @@ def forward_context(self) -> Generator[None, None, None]:
             yield
 
     def state_dict(self) -> Dict[str, Any]:
+        """``NativeMixedPrecisionPlugin.on_save_checkpoint`` is deprecated in v1.6.
+
+        Lightning will auto-save NativeMixedPrecisionPlugin state with ``NativeMixedPrecisionPlugin.state_dict`` instead
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/native_amp.py:114]:
```suggestion
        Lightning will auto-save NativeMixedPrecisionPlugin state with ``NativeMixedPrecisionPlugin.state_dict`` instead.
```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 3]
```diff
@@ -94,20 +94,17 @@ def optimizer_step(
...
+
         return amp.state_dict()
 
     def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
-        amp.load_state_dict(state_dict)
-
-    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         """``ApexMixedPrecisionPlugin.on_load_checkpoint`` is deprecated in v1.6.
 
         Lightning will auto-restore ApexMixedPrecisionPlugin state with ``ApexMixedPrecisionPlugin.load_state_dict``
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/apex_amp.py:107]:
```suggestion
        Lightning will automatically restore ``ApexMixedPrecisionPlugin``'s state with ``ApexMixedPrecisionPlugin.load_state_dict``
```

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 4]
```diff
@@ -109,23 +109,20 @@ def forward_context(self) -> Generator[None, None, None]:
             yield
 
     def state_dict(self) -> Dict[str, Any]:
+        """``NativeMixedPrecisionPlugin.on_save_checkpoint`` is deprecated in v1.6.
+
+        Lightning will auto-save NativeMixedPrecisionPlugin state with ``NativeMixedPrecisionPlugin.state_dict``
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/native_amp.py:114]:
```suggestion
        Lightning will automatically save ``NativeMixedPrecisionPlugin``'s state with ``NativeMixedPrecisionPlugin.state_dict``
```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 5]
```diff
@@ -109,23 +109,20 @@ def forward_context(self) -> Generator[None, None, None]:
             yield
 
     def state_dict(self) -> Dict[str, Any]:
+        """``NativeMixedPrecisionPlugin.on_save_checkpoint`` is deprecated in v1.6.
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/native_amp.py:112]: the docstrings are in the wrong method. did you mean to implement on_save/load_checkpoint in this class, and then call super().on_save_checkpoint / super().on_load_checkpoint ?

Comment (jjenniferdai) [reply pytorch_lightning/plugins/precision/native_amp.py:112]: we made them empty [in Stateful PR discussion](https://github.com/PyTorchLightning/pytorch-lightning/pull/11638#discussion_r799107580), and here I moved the docstring to avoid deprecation warnings - open to suggestions

Comment (ananthsub) [reply pytorch_lightning/plugins/precision/native_amp.py:112]: ah okay

Comment (carmocca) [reply pytorch_lightning/plugins/precision/native_amp.py:112]: I think this is a very confusing choice.

Why don't we set the base class empty method to do


```py
f"""``{self.__class__.__name__}.on_save_checkpoint``is deprecated...
...
"""
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -430,6 +430,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `LightningLoggerBase.agg_and_log_metrics` in favor of `LightningLoggerBase.log_metrics` ([#11832](https://github.com/PyTorchLightning/pytorch-lightning/pull/11832))
 
 
+- Deprecated `PrecisionPlugin.on_save/load_checkpoint` in favor of `state_dict/load_state_dict` ([#11978](https://github.com/PyTorchLightning/pytorch-lightning/pull/11978))
```
Comment (awaelchli) [inline CHANGELOG.md:433]:
```suggestion
- Deprecated `PrecisionPlugin.on_{save,load}_checkpoint` in favor of `PrecisionPlugin.{state_dict,load_state_dict}` ([#11978](https://github.com/PyTorchLightning/pytorch-lightning/pull/11978))
```

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 7]
```diff
@@ -259,3 +259,15 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
             state_dict: the precision plugin state returned by ``state_dict``.
         """
         pass
+
+    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/precision_plugin.py:263]: I would also be fine dropping it completely

**Review by awaelchli — APPROVED** [review]

File: tests/deprecated_api/test_remove_1-8.py [hunk 8]
```diff
@@ -608,3 +609,31 @@ def on_pretrain_routine_end(self, trainer, pl_module):
         match="The `Callback.on_pretrain_routine_end` hook has been deprecated in v1.6" " and will be removed in v1.8"
     ):
         trainer.fit(model)
+
+
+def test_v1_8_0_precplugin_checkpointhooks(tmpdir):
```
Comment (awaelchli) [inline tests/deprecated_api/test_remove_1-8.py:614]:
```suggestion
def test_v1_8_0_precision_plugin_checkpoint_hooks(tmpdir):
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 9]
```diff
@@ -363,3 +366,16 @@ def _check_deprecated_callback_hooks(trainer: "pl.Trainer") -> None:
                     f"The `Callback.{hook}` hook has been deprecated in v1.6 and"
                     f" will be removed in v1.8. Please use `Callback.on_fit_start` instead."
                 )
+
+
+def _check_precplugin_checkpoint_hooks(trainer: "pl.Trainer") -> None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:371]:
```suggestion
def _check_precision_plugin_checkpoint_hooks(trainer: "pl.Trainer") -> None:
```

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 7 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/precision/precision_plugin.py:267]: I also don't have a strong preference on keeping this or not

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
